lib/util.js

```diff
@@ -738,7 +738,9 @@ function formatSet(ctx, value, recurseTimes, visibleKeys, keys) {
   var output = [];
   value.forEach(function(v) {
     var nextRecurseTimes = recurseTimes === null ? null : recurseTimes - 1;
-    var str = formatValue(ctx, v, nextRecurseTimes);
+    var str = ctx.seen.includes(v) ?
+      ctx.stylize('[Circular]', 'special') :
+      formatValue(ctx, v, nextRecurseTimes);
```

Comment:
I am aware of that (ref https://github.com/nodejs/node/pull/14492#discussion_r130090984) but I would rather keep the concerns separated.
And I am actually working on improving the inspect performance in general at the moment and I think it would be best to keep the performance changes in that one but I did not yet have enough time to finish it.